### PR TITLE
Feature/storage/upload stream optimizations

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Previously, all partitioned uploads buffered each block before uploading
it. Now, some partitioned uploads buffer while some window the base
stream. Decision making is outlined in order below.

1. If the stream length is unknown: buffered upload. (necessary for
   providing content length at REST layer)
2. Else, if stream length is under defined threshold: one-shot upload.
3. Else, if sync call: windowed stream. (Parallel upload only offered in async
   context; no point of buffering if sequential upload)
4. Else, if max concurrency == 1: windowed stream. (again, not parallel
   upload)
5. Else, buffered upload.

### CODE REVIEW

**Comment by jaschrep-msft** [comment]
I'll be starting some basic benchmarking for this change as compared to last public release, measuring time and memory use. Note this means we'll also be testing changes of jumbo-blob support.

Will be measuring small, medium, and large files across sync and async contexts, with max concurrencies of 1 and 2. The belief is that this will save significant memory and some time in sequential uploads, with minimal effect on parallel uploads. These beneficial effects should be more noticeable with large partition sizes.

Will move to active PR upon acceptable results of this testing.

**Comment by jaschrep-msft** [comment]
The performance differences of these PartitionedUploader changes compared to our latest public release are listed below. Previous tests of latest public release to jumbo-blob support this is built on showed no significant changes. We tested paths where the uploader used Put Blob, buffered parallel upload, and a non-buffered sequential upload. Buffered sequential upload was not tested and there is no non-buffered parallel upload. Sample data sizes are large enough such that any retries will not significantly throw off results.

**Put blob**
No changes in time, CPU intensity, or memory usage.

**Buffered parallel upload**
Most configurations run a little faster. CPU intensity and memory usage rise in step. This was unexpected, as these changes focused on the next section.

**Non-buffered sequential upload**
Major improvements in most configurations, organized by block size. Memory usage was flat across all block sizes (barebones process to upload a directory used 53MB), whereas before it went up by a large coefficient based on block size.
* 4MB blocks: 10-20% speed improvement depending on other settings and overall file size. CPU intensity rises with this and memory usage goes down ~5%.
* 8MB blocks: speed and CPU inconsistent across sync and async contexts. Async showed similar results to 4MB blocks but sync showed a minor slowdown. CPU was in step with speed, and memory usage went down nearly 30% across both sync and async.
* 32MB blocks: 15% speed improvement with a 40% CPU intensity increase. Memory usage went down by 70%
* 100MB blocks: 40% speed improvement with 50% CPU intensity increase. Memory usage went down by almost 90%.

File: sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs:9]: Since WindowStream is now abstract I think we could get rid of this type and just use WindowStream instead where AbsolutePolistion is needed. 
Btw. "StreamSlice" and "WindowStream" seem to be same thing conceptually (naming wise).

File: sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs:25]: So in prod code we decided to use inheritance to model seekability vs non-seekability (SeekableWindowStream and UnseekableWindowStream). But in tests we drive this through constructor flag (i.e. we don't have SeekablePredicatableStream). 

I was about to make a comment that we could just accept flag `seekable` in WindowStream instead of creating so many types, but I thought it's gonna be lot of if/else branching logic. But then I see this code and I think it's more concise than inheritance based solution. If you don't mind I'd try to make WindowStream accept a flag and see if it would look better (less complex).

**Review by kasobol-msft — COMMENTED** [review]
Overall looks good.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:450]: This was previously existing logic that was moved (and renamed to reflect linq naming patterns). We will throw elsewhere if the stream is null, but I think what amounts to a safe length reader should absolutely never throw.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:462]: I mean, someone can write code that can do that. `Stream` documentation is very explicit that length, position, and seekability are an all-or-nothing package, and so `CanSeek` is the indicator. Anyone who gives us a stream that behaves otherwise is breaking the .NET API contract.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:485]: It seems there is no point in handwritten code where we validate user-specified block size > 0. That would likely cause failure at some other layer, and would be an existing bug in today's GA code (this logic previously existed and was moved in refactors). We can probably take the initial fetch of block size and assert it's > 0, and then clamp `acceptableBlockSize` to also be > 0. While it may change how we fail in such a scenario, we'll still fail in some regard. Additionally, it's extremely unlikely customers are currently passing in values <= 1 in production code, so such a change should not pose a risk to anyone.

File: sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs:9]: I was discussing this with Sean the other day regarding `PooledMemoryStream` and his use of it for OpenWrite. Unfortunately, since both this and his PRs touch the same code, it would not be fun for either of us to make the appropriate refactors until we've both merged. #12820 was filed to address this regarding `PooledMemoryStream`, and we can address `WindowStream` along with it.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:450]: I'd rather define contract of this method (via docs) that it requires non-null argument and throw if someone supplies null. I think attempt to measure length of null stream is like trying to divide by 0.

In general it's good to add good validation in publicly facing entry points and then assume input correctness in private parts. Such approach reduces cyclomatic complexity of the code.

Some more reading https://enterprisecraftsmanship.com/posts/defensive-programming/

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:462]: Yeah. I'm assuming Stream implemented according to .NET guidance. Thank you!

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:485]: I'd add such validation then and let it fail fast and clean.

File: sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs:25]: I chose to split the implementations because there were differences in managing state instead of just when to pass things through. `UnseekableWindowStream` has to maintain bytes read and an overall max possible bytes before finishing the stream facade. These are useless in `SeekableWindowStream`, which operates on public `Length` properties and has no use for a position. Perhaps we want to try a composition approach, where `WindowStream` implements what it currently does and contains some implementation of the rest of these methods? I still think the if/else across all these methods would be a bit much to read, especially when they're all branching off the same condition.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs:25]: Yeah, I can imagine this being too much of spaghetti with these ifs. Your call.

File: sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StreamSlice.cs:9]: Ok.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:485]: Was incorrect about how users can set block size. They can set whatever they want, but if they set a non-positive number then we ignore it and use our defaults. So at this point, `blockSize` will always be positive and there is no need to insert a fail-fast check. Will add `Min(1, blockSize / 2)` to this line in case someone passes block size of 1, but that's more of a sanity check.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs:450]: Added argument null check to the overall upload method on this class.

**Review by seanmcc-msft — APPROVED** [review]

**Pull request merged**
